src/Json.php

```diff
@@ -165,84 +165,107 @@ protected static function recursiveJsonExprFinder(
      */
     public static function prettyPrint($json, array $options = [])
     {
-        $tokens       = preg_split('|([\{\}\]\[,])|', $json, -1, PREG_SPLIT_DELIM_CAPTURE);
-        $result       = '';
-        $indentLevel  = 0;
         $indentString = isset($options['indent']) ? $options['indent'] : '    ';
-        $inLiteral    = false;
 
-        $openingBrackets = ['{', '['];
-        $closingBrackets = ['}', ']'];
-
-        $bracketPairs = array_combine(
-            $openingBrackets,
-            $closingBrackets
-        );
-
-        $count = count($tokens);
-
-        for ($i = 0; $i < $count; ++$i) {
-            $token = trim($tokens[$i]);
-
-            if ($token === '') {
-                continue;
+        $json = trim($json);
+        $length = strlen($json);
+        $stack = [];
+
+        $result = '';
+        $inLiteral = false;
+
+        for ($i = 0; $i < $length; ++$i) {
+            switch ($json[$i]) {
+                case '{':
+                case '[':
+                    if (! $inLiteral) {
+                        $stack[] = $json[$i];
+
+                        $result .= $json[$i];
+                        while (isset($json[$i + 1]) && preg_match('/\s/', $json[$i + 1])) {
+                            ++$i;
+                        }
+                        if (isset($json[$i + 1]) && $json[$i + 1] !== '}' && $json[$i + 1] !== ']') {
+                            $result .= "\n" . str_repeat($indentString, count($stack));
+                        }
+                        continue 2;
+                    }
+                    break;
```

Comment:
I'd break early, so that you can remove a level of indentation:

```php
if ($inLiteral) {
    break;
}

$stack[] = $json[$i];

$result .= $json[$i];
while (isset($json[$i + 1]) && preg_match('/\s/', $json[$i+1])) {
    ++$i;
}
if (isset($json[$i + 1]) && $json[$i + 1] !== '}' && $json[$i + 1] !== ']') {
    $result .= "\n" . str_repeat($indentString, count($stack));
}
continue 2;
```

I'd do that in each of the case statements below as well.

Reply:
Updated, thanks!